training/tensorboard/pytorch/pytorch_smdataparallel_mnist_demo.ipynb

```diff
@@ -0,0 +1,811 @@
+    "\n",
+    "### Customizing training script for TensorBoard data collection\n",
+    "\n",
+    "In order to activate TensorBoard data collection, you need to explicitly modify the training script to store the model data using the `torch.utils.tensorboard.SummaryWriter` class. Note that this requires TensorBoard installed in the training container, so you need to add the `tensorboard` dependency in `requirements.txt`.\n",
+    "\n",
+    "When collecting the model tensors from a distributed training job, using the `SummaryWriter` across multiple ranks can cause a duplication of data collection, so make sure to specify the rank for the `SummaryWriter`. The following code snippet shows the necessary changes to set up the SummaryWriter class and collect model tensors.\n",
+    "\n",
+    "\n",
+    "```\n",
+    "from torch.utils.tensorboard import SummaryWriter\n",
+    "\n",
+    "...\n",
+    "args.rank = rank = dist.get_rank()\n",
+    "args.local_rank = local_rank = int(os.getenv(\"LOCAL_RANK\", -1))\n",
+    "...\n",
+    "writer = SummaryWriter('/opt/ml/output/tensorboard') if rank == 0 and local_rank == 0 else None\n",
+    "...\n",
+    "\n",
+    "def train(args, model, device, train_loader, optimizer, epoch):\n",
+    "    ...\n",
+    "    # Only write on rank 0 to avoid duplicate entries\n",
+    "    if writer:\n",
+    "        writer.add_scalar(\"example_metric\", example_metric)\n",
+    "        \n",
+    "# Close writer at the end of training \n",
+    "writer.close()\n",
+    "```\n",
+    "To see the full training script, run the cell below:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "tags": []
+   },
+   "outputs": [],
+   "source": [
+    "!pygmentize code/train_pytorch_smdataparallel_mnist.py"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Estimator function options\n",
+    "\n",
+    "To instruct SageMaker to find the local path where the model tensors are saved and upload the tensor data to an Amazon S3 bucket, configure the `sagemaker.debugger.TensorBoardOutputConfig` class as follows.\n"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from sagemaker.debugger import TensorBoardOutputConfig\n",
+    "\n",
+    "tensorboard_s3_output_path = \"s3://{}/{}\".format(bucket, prefix)\n",
+    "\n",
+    "# Create a TensorBoardOutputConfig object to configure automatic data upload from\n",
+    "# the training job's local data storage to S3\n",
+    "tensorboard_output_config = TensorBoardOutputConfig(\n",
+    "    s3_output_path=tensorboard_s3_output_path,\n",
+    "    container_local_output_path=\"/opt/ml/output/tensorboard\",\n",
+    ")"
+   ]
+  },
+  {
```

Comment:
```suggestion
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The following two code cells show how to set the `tensorboard_output_config` parameter and other necessary parameters for the SageMaker PyTorch estimator class, and launch a training job by running the `estimator.fit()` method. Also note that you need to pass the `requirements.txt` file to the `SAGEMAKER_REQUIREMENTS` environment variable through the estimator. SageMaker checks the requirements and installs packages (the `tensorboard` package in this case) listed in the file.\n"
   ]
  },
  {
```

Reply:
After incorporating this suggestion, would you please check if the parentheses are closing correctly? I see that the closing parentheses are highlighted in my suggestions, and worried if I made any syntax error.